src/irr/asset/CGraphicsPipelineLoaderMTL.cpp

```diff
@@ -288,28 +327,73 @@ void main()
         else
             Ns = PC.params.Ns;
 
-        color = Ka*Ia + Kd*NdotL;
+        vec3 diff = Kd*params.NdotL;
         switch (PC.params.extra&ILLUM_MODEL_MASK)
         {
+        case 1:
+            color = Ia*Ka + diff;
+            break;
         case 2:
-        case 3://2 + reflection map
-        case 4://3 with transparency (glass)
-        case 6:
-        case 8://reflection map
-        case 9://reflection map
-            color += Ks*pow(VdotR, Ns);
+        case 3://2 + IBL
+        case 5://basically same as 3
+        case 8://basically same as 5
+        {
+            float fr = Fresnel_dielectric(PC.params.Ni, params.NdotL);
+            vec3 spec = Ks*pow(params.NdotH, Ns);
+            color += Ia*Ka + mix(diff, spec, fr);
+        }
             break;
-        case 5:
+        case 4:
+        case 6:
         case 7:
-            color += Ks*pow(VdotR, Ns)*Fresnel_dielectric(PC.params.Ni, NdotL);
+        case 9://basically same as 4
+        {
+            float fr = Fresnel_dielectric(PC.params.Ni, params.NdotL);
+            vec3 spec = Ks*pow(params.NdotH, Ns);
+            color = fr*(Ia*Ka+spec);
+        }
             break;
         default:
             break;
         }
     }
     else color = Kd;
 
-    OutColor = vec4(color, d);
+    return color;  
+}
+
+#endif //_IRR_BSDF_COS_EVAL_DEFINED_
+
+
+
+#ifndef _IRR_FRAG_MAIN_DEFINED_
+#define _IRR_FRAG_MAIN_DEFINED_
+
+void main()
+{
+    irr_glsl_BSDFIsotropicParams bsdfParams = irr_glsl_computeLighting();
+    vec3 color = irr_glsl_bsdf_cos_eval(bsdfParams);
+
+    float d = PC.params.d;
+
+    //another illum model switch, required for illum=4,6,7,9 to compute alpha from fresnel (taken from opacity map or constant otherwise)
+    switch (PC.params.extra&ILLUM_MODEL_MASK)
+    {
+    case 4:
+    case 6:
+    case 7:
+    case 9:
+        d = Fresnel_dielectric(PC.params.Ni, bsdfParams.NdotL);
+        break;
+    default:
+        if ((PC.params.extra&(map_d_MASK)) == (map_d_MASK))
+            d = texture(map_d, UV).r;
+        break;
+    }
+#define Intensity 1000.0
+    // `/dot(ViewPos,Viewpos)` conflicts with requirement to be able to override lighting just by redefinition of irr_glsl_computeLighting()
+    // ^^^ TODO, decision to make
```

Comment:
you shouldn't be evaluating the lighting (even the attentuation or intensity) inside the `main` function, all lights need to be wholly evaluated inside `irr_bsdf_cos_eval`.

Reply:
hm, then i'd say we shouldnt normalize `L` vector in `irr_glsl_calcBSDFIsotropicParams()` because otherwise `irr_bsdf_cos_eval()` has no way to take distance from light

Reply:
or better: normalize but add field `invlenL2` to structure

Reply:
but you know the `L` you've passed to irr_glsl_calcBSDFIsotropicParams ;) thats not getting modified.

Reply:
yes, but i dont know it inside `bsdf_cos_eval`. If u want attenuation factor in `irr_bsdf_cos_eval` then length of L must be known in there. As for now i made `invlenL2` field in the struct, will commit in up to 1h

Reply:
attenuation of a light is not part of a BRDF/BSDF, I thought that was clear?

Reply:
yea, I know, turns out I misunderstood your requirements

Reply:
`brdf_cos == BRDF()*cos(theta)`

BRDF only has the material's probability of reflecting incoming photons in a given direction

cos(theta)/LightDistance2 is the solidAngle factor for a point light (how small it appears at a given angle and distance - its an oxymoron cause point light doesn't have a size, but for a small disk that's what it would be)

Reply:
area lights, line lights, environment maps, etc. they have different solidAngle factors (albeit all include `cos(theta)` thats why we put it in `brdf_cos` functions)